### PR TITLE
[MIG] Master stock account new api dut

### PR DESCRIPTION
**Pull request by dut-odoo**
<strong> Task: </strong> https://www.odoo.com/web#id=13742&view_type=form&model=project.task&menu_id=3942&action=327
<strong> Pad: </strong> https://pad.odoo.com/p/openerp-project.task-0PPOGPR9U7

### CODE REVIEW

File: addons/purchase/test/average_price.yml [hunk 1]
```diff
@@ -10,8 +10,8 @@
     uom_po_id: product.product_uom_kgm
     valuation: real_time
     cost_method: average
-    property_stock_account_input: o_expense
-    property_stock_account_output: o_income
+    property_stock_account_input_id: o_expense
```
Comment (hmo-odoo) [inline addons/purchase/test/average_price.yml]: make separate commit for field change

File: addons/stock_account/models/account_chart_template.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    _inherit = "account.chart.template"
+
+    @api.model
+    def generate_journals(self, acc_template_ref, company, journals_dict=None):
+        journal_to_add = [{'name': _('Stock Journal'), 'type': 'general', 'code': 'STJ', 'favorite': False, 'sequence': 8}]
+        super(AccountChartTemplate, self).generate_journals(acc_template_ref=acc_template_ref, company=company, journals_dict=journal_to_add)
+
+    @api.multi
+    def generate_properties(self, acc_template_ref, company, property_list=None):
+        self.ensure_one()
```
Comment (hmo-odoo) [inline addons/stock_account/models/account_chart_template.py]: remove ensure_one()

File: addons/stock_account/models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+
+class AccountInvoice(models.Model):
+    _inherit = "account.invoice"
+
+    @api.model
+    def invoice_line_move_line_get(self):
+        res = super(AccountInvoice,self).invoice_line_move_line_get()
+        if self.company_id.anglo_saxon_accounting:
+            if self.type in ('out_invoice','out_refund'):
```
Comment (jir-odoo) [inline addons/stock_account/models/account_invoice.py]: Use self.company_id.anglo_saxon_accounting and self.type in ('out_invoice','out_refund')

File: addons/stock_account/models/product_template.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+
+class ProductTemplate(models.Model):
+    _name = 'product.template'
+    _inherit = 'product.template'
+
+    property_valuation = fields.Selection([
+        ('manual_periodic', 'Periodic (manual)'),
+        ('real_time', 'Perpetual (automated)')],
+        string='Inventory Valuation', copy=True,
```
Comment (jir-odoo) [inline addons/stock_account/models/product_template.py]: No need copy true

File: addons/stock_account/models/account_chart_template.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        for record in todo_list:
+            if record == 'property_stock_journal_id':
+                relation = 'account.journal'
+            else:
+                relation = 'account.account'
+                account = getattr(self, record)
+                journal = account and acc_template_ref.get(account.id)
+
+            if journal:
+                field = IrModelFields.search([('name', '=', record), ('model', '=', 'product.category'), ('relation', '=', relation)], limit=1)
```
Comment (WHPODOO) [inline addons/stock_account/models/account_chart_template.py]: IrModelFields  use once then put here directly like : 
```
field = self.env['ir.model.fields'].search([('name', '=', record), ('model', '=', 'product.category'), ('relation', '=', relation)], limit=1)
```
:turtle:

File: addons/stock_account/models/product_template.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    def _get_product_accounts(self):
+        """ Add the stock accounts related to product to the result of super()
+        @return: dictionary which contains information regarding stock accounts and super (income+expense accounts)
+        """
+        accounts = super(ProductTemplate, self)._get_product_accounts()
+        accounts.update({
+            'stock_input': self.property_stock_account_input_id or self.categ_id.property_stock_account_input_categ_id,
+            'stock_output': self.property_stock_account_output_id or self.categ_id.property_stock_account_output_categ_id,
+            'stock_valuation': self.categ_id.property_stock_valuation_account_id or False,
+        })
```
Comment (WHPODOO) [inline addons/stock_account/models/product_template.py]: You also write above code like  : 
```
accounts[ 'stock_input'] = self.property_stock_account_input_id or self.categ_id.property_stock_account_input_categ_id
accounts[ 'stock_output'] = self.property_stock_account_output_id or self.categ_id.property_stock_account_output_categ_id
accounts[ 'stock_valuation'] = self.categ_id.property_stock_valuation_account_id or False
```
I think its looks more readable code :turtle:

File: addons/stock_account/models/stock_history.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    price_unit_on_quant = fields.Float('Value')
+    inventory_value=  fields.Float(compute='_get_inventory_value', string="Inventory Value", readonly=True)
+    source = fields.Char()
+    product_template_id = fields.Many2one('product.template', 'Product Template', required=True)
+    serial_number = fields.Char('Serial Number', required=True)
+
+    @api.model
+    def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
+        res = super(StockHistory, self).read_group(domain, fields, groupby, offset=offset, limit=limit, orderby=orderby, lazy=lazy)
+        Stock_History = self.env['stock.history']
```
Comment (WHPODOO) [inline addons/stock_account/models/stock_history.py]: when you assign value of object then variable name should be camel case not snake case like : 
```
StockHistory = self.env['stock.history']
```

File: addons/stock_account/models/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    _inherit = "account.invoice.line"
+
+    def _get_anglo_saxon_price_unit(self):
+        self.ensure_one()
+        return self.product_id.standard_price
+
+    def _get_price(self, invoice, company_currency_id, price_unit):
+        self.ensure_one()
+        if invoice.currency_id.id != company_currency_id:
+            price = self.env['res.currency'].with_context({'date': invoice.date_invoice}).compute(company_currency_id, invoice.currency_id.id, price_unit * self.quantity)
```
Comment (mdi-odoo) [inline addons/stock_account/models/account_invoice.py]: Call the method using recordset(s) .

File: addons/stock_account/models/stock_quant.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, models
+
+
+class StockQuant(models.Model):
+    _inherit = "stock.quant"
+
+    @api.model
+    def _get_inventory_value(self, quant):
```
Comment (mdi-odoo) [inline addons/stock_account/models/stock_quant.py]: Why @api.model ? Here 'quant' is a recordset(s) .

File: addons/stock_account/models/stock_quant.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+        ''' This function is called at the end of negative quant reconciliation and does the accounting entries adjustemnts and the update of the product cost price if needed
+        '''
+        super(StockQuant, self)._price_update(newprice)
+        AccountMove = self.env['account.move']
+        for quant in self:
+            move = self._get_latest_move(quant)
+            valuation_update = newprice - quant.cost
+            # this is where we post accounting entries for adjustment, if needed
+            # If neg quant period already closed (likely with manual valuation), skip update
+            if not quant.company_id.currency_id.is_zero(valuation_update) and AccountMove.browse(move.id)._check_lock_date():
```
Comment (mdi-odoo) [inline addons/stock_account/models/stock_quant.py]: Why you need to rebrowse? Here 'move' is already a browse record .

File: addons/stock_account/models/stock_quant.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    @api.model
+    def _quant_create(self, qty, move, lot_id=False, owner_id=False, src_package_id=False, dest_package_id=False, force_location_from=False, force_location_to=False):
+        quant = super(StockQuant, self)._quant_create(qty, move, lot_id=lot_id, owner_id=owner_id, src_package_id=src_package_id, dest_package_id=dest_package_id, force_location_from=force_location_from, force_location_to=force_location_to)
+        quant._account_entry_move(move)
+        return quant
+
+    @api.model
+    def move_quants_write(self, quants, move, location_dest_id, dest_package_id, lot_id=False, entire_pack=False):
+        res = super(StockQuant, self).move_quants_write(quants, move, location_dest_id,  dest_package_id, lot_id=lot_id, entire_pack=entire_pack)
+        Stock_Quant = self.env['stock.quant']
```
Comment (mdi-odoo) [inline addons/stock_account/models/stock_quant.py]: Rename the variable 'Stock_Quant' to 'StockQuant' .

File: addons/stock_account/models/stock_quant.py [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+        for quant in quants:
+            Stock_Quant += quant
+        Stock_Quant._account_entry_move(move)
```
Comment (mdi-odoo) [inline addons/stock_account/models/stock_quant.py]: Why you need this stuff ? You can directly call method on 'quants'.

File: addons/stock_account/models/account_invoice.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+
+    def _get_anglo_saxon_price_unit(self):
+        self.ensure_one()
+        return self.product_id.standard_price
+
+    def _get_price(self, invoice, company_currency, price_unit):
+        self.ensure_one()
```
Comment (mdi-odoo) [inline addons/stock_account/models/account_invoice.py]: No need of self.ensure_one() here .

File: addons/stock_account/models/product_category.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class ProductCategory(models.Model):
+    _inherit = 'product.category'
+
+    property_valuation = fields.Selection([
+        ('manual_periodic', 'Periodic (manual)'),
+        ('real_time', 'Perpetual (automated)')],
+        string='Inventory Valuation',
+        required=True, copy=True,
```
Comment (mdi-odoo) [inline addons/stock_account/models/product_category.py]: The value of 'copy' defaults to True for all fields except computed fields and one2many fields .

File: addons/stock_account/models/product_category.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+             "Method'. The inventory variation account set on the product "
+             "category will represent the current inventory value, and the "
+             "stock input and stock output account will hold the counterpart "
+             "moves for incoming and outgoing products.")
+    property_cost_method = fields.Selection([
+        ('standard', 'Standard Price'),
+        ('average', 'Average Price'),
+        ('real', 'Real Price')],
+        string="Costing Method",
+        required=True, copy=True,
```
Comment (mdi-odoo) [inline addons/stock_account/models/product_category.py]: Same as above.

File: addons/stock_account/models/product_category.py (hunk 14 above)
Comment (dut-odoo) [inline addons/stock_account/models/product_category.py]: @mdi-odoo,
Yes you are right default copy=True is only for normal field and for other relational field it is False
but here "property_valuation" is property field so that’s why i have to set copy=True.

File: addons/stock_account/models/stock_history.py [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+
+    def _get_inventory_value(self):
+        history_date = self.env.context.get('history_date')
+        Product = self.env["product.product"]
+        for line in self:
+            if line.product_id.cost_method == 'real':
+                line.inventory_value = line.quantity * line.price_unit_on_quant
+            else:
+                line.inventory_value = line.quantity * Product.get_history_price(line.product_id.id, line.company_id.id, date=history_date)
+
```
Comment (mdi-odoo) [inline addons/stock_account/models/stock_history.py]: Call the method using recordset(s) .

File: addons/stock_account/models/stock_quant.py (hunk 10 above)
Comment (dut-odoo) [inline addons/stock_account/models/stock_quant.py]: @mdi-odoo,
because "move" is a <b>stock.move</b> and here we need <b>account.move</b>  for calling _check_lock_date method.

File: addons/stock_account/models/stock_history.py [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    inventory_value=  fields.Float(compute='_get_inventory_value', string="Inventory Value", readonly=True)
+    source = fields.Char()
+    product_template_id = fields.Many2one('product.template', 'Product Template', required=True)
+    serial_number = fields.Char('Serial Number', required=True)
+
+    @api.model
+    def read_group(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
+        res = super(StockHistory, self).read_group(domain, fields, groupby, offset=offset, limit=limit, orderby=orderby, lazy=lazy)
+        StockHistories = self.env['stock.history']
+        history_date = self.env.context.get('history_date', Datetime.now())
```
Comment (hmo-odoo) [inline addons/stock_account/models/stock_history.py]: can you use fields.datetime.today() instead of Datetime.now() ?

File: addons/stock_account/models/stock_quant.py [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+
+    def _account_entry_move(self, move):
+        """
+        Accounting Valuation Entries
+
+        move: Move to use. browse record
+        """
+        StockLocation = self.env['stock.location']
+        location_from = move.location_id
+        location_to = self[0].location_id
```
Comment (hmo-odoo) [inline addons/stock_account/models/stock_quant.py]: why self[0] ?

File: addons/stock_account/wizard/stock_change_standard_price.py [hunk 19]
```diff
@@ -1,71 +1,44 @@
...
-            prod_obj = self.pool.get('product.product')
-            rec_id = prod_obj.browse(cr, uid, rec_id, context=context).product_tmpl_id.id
-        prod_obj = self.pool.get('product.template')
-
-        res = self.browse(cr, uid, ids, context=context)[0]
+        if self.env.context.get("active_model") == 'product.product':
+            rec_id = self.env['product.product'].browse(rec_id).product_tmpl_id.id
 
-        prod_obj.do_change_standard_price(cr, uid, [rec_id], res.new_price, res.counterpart_account_id.id, context)
+        self.env['product.template'].browse(rec_id).do_change_standard_price(self.new_price, self.counterpart_account_id.id)
```
Comment (hmo-odoo) [inline addons/stock_account/wizard/stock_change_standard_price.py]: no need browse record of product.template.

File: addons/stock_account/models/stock_quant.py (hunk 18 above)
Comment (dut-odoo) [inline addons/stock_account/models/stock_quant.py]: @hmo-odoo, 
I used self[0] as per old code requirement.
Old Code: https://github.com/odoo/odoo/blob/master/addons/stock_account/stock_account.py#L163-L174

**Comment by hmo-odoo** [comment]
continue work on this PR: https://github.com/odoo/odoo/pull/12525
